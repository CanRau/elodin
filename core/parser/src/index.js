import Parser from './Parser'
import parse from './parse'
import errorTypes from './errorTypes'

export { Parser, parse, errorTypes }
