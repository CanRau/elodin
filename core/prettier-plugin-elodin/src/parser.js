import { format } from '@elodin/format'

export default function parse(text, _, options) {
  const formatted = format(text)

  const errors = formatted.errors
  let formattedText = formatted.code

  if (errors.length > 0) {
    let formattedText = text
  }

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    error: errors[0],
    end: text.length,
    source: text,
    start: 0,
  }
}
