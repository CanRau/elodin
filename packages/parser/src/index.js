import Parser from './Parser'
import parse from './parse'

export { Parser, parse }
