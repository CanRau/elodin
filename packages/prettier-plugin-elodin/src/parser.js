import { format } from '@elodin/format'

export default function parse(text) {
  const formattedText = format(text)

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    end: text.length,
    source: text,
    start: 0,
  }
}
