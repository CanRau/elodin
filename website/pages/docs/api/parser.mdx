import { make as DocLayout } from '../../../components/DocLayout.bs.js'

export default DocLayout

# Parser

`@elodin/parser`

The parser turns a string of Elodin code into an AST (Abstract Syntax Tree).

## parse

It exposes a `parse` method that just takes an single string of code.

### Example

```javascript
import { parse } from '@elodin/parser'

const input = `
view Button {
  backgroundColor: red
  paddingLeft: 10
  paddingRight: 10
}
`

const ast = parse(input)
```

## AST Nodes

The AST consists of the following AST nodes:

### File

| Property | Type | Description |
| -------- | ---- | ----------- |
| body     |      |             |

### Style

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| format   |      |             |
| body     |      |             |
| comments |      |             |

### Variant

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| body     |      |             |
| comments |      |             |

### Conditional

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| operator |      |             |
| boolean  |      |             |
| body     |      |             |

### Declaration

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| dynamic  |      |             |
| raw      |      |             |

### FunctionExpression

| Property | Type | Description |
| -------- | ---- | ----------- |
| callee   |      |             |
| params   |      |             |

### Float

| Property   | Type | Description |
| ---------- | ---- | ----------- |
| integer    |      |             |
| fractional |      |             |
| negative   |      |             |

### Integer

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
| negative |      |             |

### Identifier

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### String

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### Variable

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
