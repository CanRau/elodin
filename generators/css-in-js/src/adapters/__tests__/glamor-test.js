import { parse } from '@elodin/parser'

import { createGenerator } from '../../index'

import glamorAdapter from '../glamor'

const file = `
variant Type {
  Primary
  Secondary
}

variant Mode {
  Dark
  Light
}

view Button {
  __border: 0
  backgroundColor: red
  paddingLeft: 10
  borderWidth: $borderWidth
  [Type=Primary] {
    backgroundColor: red
    paddingLeft: $paddingLeft

    [@hover] {
      paddingLeft: 10
      paddingRight: $paddingRight
    }
    [Mode=Dark] {
      backgroundColor: blue
      paddingLeft: $padLeft
    }
  }
  [Mode=Light] {
    backgroundColor: green
  }
}

text Label {
  lineHeight: $lineHeight

  [@hover] {
    color: red
    fontSize: $fontSize
  }

  [@viewportWidth>=320] {
    color: green
    fontSize: $mediaFontSize
    [@hover] {
      fontSize: $mediaHoverFontSize
      color: blue
    }
  }
}`

describe('Compiling to CSS and JavaScript using glamor', () => {
  it('should return a map of files', () => {
    const { ast } = parse(file)

    expect(
      createGenerator({
        adapter: glamorAdapter(),
      })(ast, 'index.elo')
    ).toMatchSnapshot()
  })

  it('should pass adapter configuration', () => {
    const { ast } = parse(file)

    expect(
      createGenerator({
        adapter: glamorAdapter({ dynamicImport: true }),
      })(ast, 'index.elo')
    ).toMatchSnapshot()
  })
})
