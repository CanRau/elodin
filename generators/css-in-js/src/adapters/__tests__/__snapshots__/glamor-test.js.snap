// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiling to CSS and JavaScript using glamor should pass adapter configuration 1`] = `
Object {
  "Button.elo.css": "._w7qnl2 {
  border: 0px;
  background-color: red;
  padding-left: 10px
}

._w7qnl2_0-0 {
  background-color: red
}

._w7qnl2_0-0:hover {
  padding-left: 10px
}

._w7qnl2_0-0_1-0 {
  background-color: blue
}

._w7qnl2_1-1 {
  background-color: green
}",
  "Button.elo.js": "import { css } from 'glamor'
import { getClassNameFromVariantMap, getDynamicStyleFromVariantMap } from '@elodin/runtime'

const variantClassNameMap = {
  \\"\\": {},
  \\"_0-0\\": {
    \\"Type\\": \\"Primary\\"
  },
  \\"_0-0_1-0\\": {
    \\"Type\\": \\"Primary\\",
    \\"Mode\\": \\"Dark\\"
  },
  \\"_1-1\\": {
    \\"Mode\\": \\"Light\\"
  }
}

const variantStyleMap = {
'__Type-Primary': ({ borderWidth, paddingLeft, paddingRight, padLeft }) => ({
  \\"paddingLeft\\":paddingLeft,
  \\":hover\\":{\\"paddingRight\\":paddingRight}
}),
'__Type-Primary__Mode-Dark': ({ borderWidth, paddingLeft, paddingRight, padLeft }) => ({
  \\"paddingLeft\\":padLeft
})
}

export function Button(props = {}) {
  import('./Button.elo.css')
  const { borderWidth, paddingLeft, paddingRight, padLeft } = props

  const className = '_elo_view ' + getClassNameFromVariantMap('_w7qnl2', variantClassNameMap, props)

  return className + \\" \\" + css({
    \\"borderWidth\\":borderWidth,
    ...getDynamicStyleFromVariantMap(variantStyleMap, props)
  })
}",
  "Label.elo.css": "._19ixzy1:hover {
  color: red
}

@media (min-width: 320px) {
  ._19ixzy1 {
    color: green
  }
}

@media (min-width: 320px) {
  ._19ixzy1:hover {
    color: blue
  }
}",
  "Label.elo.js": "import { css } from 'glamor'

export function Label(props = {}) {
  import('./Label.elo.css')
  const { lineHeight, fontSize, mediaFontSize, mediaHoverFontSize } = props

  const className = '_elo_text _19ixzy1'

  return className + \\" \\" + css({
    \\"lineHeight\\":lineHeight,
    \\":hover\\":{\\"fontSize\\":fontSize},
    \\"@media (min-width: 320px)\\":{\\"fontSize\\":mediaFontSize,
\\":hover\\":{\\"fontSize\\":mediaHoverFontSize}}
  })
}",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
  "index.elo.js": "import './_reset.elo.css'
import { Button } from './Button.elo.js'
import { Label } from './Label.elo.js'

export {
  Button,
  Label
}",
}
`;

exports[`Compiling to CSS and JavaScript using glamor should return a map of files 1`] = `
Object {
  "Button.elo.css": "._w7qnl2 {
  border: 0px;
  background-color: red;
  padding-left: 10px
}

._w7qnl2_0-0 {
  background-color: red
}

._w7qnl2_0-0:hover {
  padding-left: 10px
}

._w7qnl2_0-0_1-0 {
  background-color: blue
}

._w7qnl2_1-1 {
  background-color: green
}",
  "Button.elo.js": "import './Button.elo.css'
import { css } from 'glamor'
import { getClassNameFromVariantMap, getDynamicStyleFromVariantMap } from '@elodin/runtime'

const variantClassNameMap = {
  \\"\\": {},
  \\"_0-0\\": {
    \\"Type\\": \\"Primary\\"
  },
  \\"_0-0_1-0\\": {
    \\"Type\\": \\"Primary\\",
    \\"Mode\\": \\"Dark\\"
  },
  \\"_1-1\\": {
    \\"Mode\\": \\"Light\\"
  }
}

const variantStyleMap = {
'__Type-Primary': ({ borderWidth, paddingLeft, paddingRight, padLeft }) => ({
  \\"paddingLeft\\":paddingLeft,
  \\":hover\\":{\\"paddingRight\\":paddingRight}
}),
'__Type-Primary__Mode-Dark': ({ borderWidth, paddingLeft, paddingRight, padLeft }) => ({
  \\"paddingLeft\\":padLeft
})
}

export function Button(props = {}) {
  const { borderWidth, paddingLeft, paddingRight, padLeft } = props

  const className = '_elo_view ' + getClassNameFromVariantMap('_w7qnl2', variantClassNameMap, props)

  return className + \\" \\" + css({
    \\"borderWidth\\":borderWidth,
    ...getDynamicStyleFromVariantMap(variantStyleMap, props)
  })
}",
  "Label.elo.css": "._19ixzy1:hover {
  color: red
}

@media (min-width: 320px) {
  ._19ixzy1 {
    color: green
  }
}

@media (min-width: 320px) {
  ._19ixzy1:hover {
    color: blue
  }
}",
  "Label.elo.js": "import './Label.elo.css'
import { css } from 'glamor'

export function Label(props = {}) {
  const { lineHeight, fontSize, mediaFontSize, mediaHoverFontSize } = props

  const className = '_elo_text _19ixzy1'

  return className + \\" \\" + css({
    \\"lineHeight\\":lineHeight,
    \\":hover\\":{\\"fontSize\\":fontSize},
    \\"@media (min-width: 320px)\\":{\\"fontSize\\":mediaFontSize,
\\":hover\\":{\\"fontSize\\":mediaHoverFontSize}}
  })
}",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
  "index.elo.js": "import './_reset.elo.css'
import { Button } from './Button.elo.js'
import { Label } from './Label.elo.js'

export {
  Button,
  Label
}",
}
`;
